Send error messages to an MQTT topic

nix-chu opened the pull request:
**Updated: 31 July 2020**

## Description
Catches errors that may occur while running Raspicam and sends it to `/v3/camera/errors` MQTT topic using new `CameraErrorHandler` class. 

Currently catches:
- Disconnected camera
- Error when loading the base canvas
- Errors when updating the data and message overlay
- Errors when loading data

If anyone has suggestions for any potential errors that may occur for the camera, please let me know! 

## Screenshots

![image](https://user-images.githubusercontent.com/58575113/88888224-4779cc00-d281-11ea-8481-85ddc474bb1a.png)

New Windows Terminal is amazing! And it's customisable 

## Steps to Test
1. Start your local MQTT broker and subscribe to `/v3/camera/errors` topic.
`mosquitto_sub -t "/v3/camera/errors"`
2. Assuming you are in Raspicam directory, run `overlay_error.py` to perform the error test.
`python overlay_error.py`

khanguslee commented on backend/picamera_backend.py:
```diff
@@ -61,13 +61,23 @@ def update_picamera_overlay(self, canvas: Canvas, layer: PiCameraOverlayLayer) -
+        try:
+            self.update_picamera_overlay(base_canvas, PiCameraOverlayLayer.base)
+        except:
+            self.send_camera_error()
```
Is there some weird python magic class/method where we can use the `self.send_camera_error()` method as the default exception handler? Just a bit annoying to wrap everything with a try except

khanguslee replied:
I knew there'd be some weird Python thing LOL

nix-chu replied:
Wow this is very intuitive, what the heck
![image](https://user-images.githubusercontent.com/58575113/88627728-22098880-d0f0-11ea-845c-971ac80dd653.png)
(This is just a random file I created btw, not the actual code lol)

I thought it would be some complicated thing, but this is as simple as it gets. Using this method, we'd have to keep using the `with` statement. Is this what you meant as the "default error handler"? @khanguslee

nix-chu replied:
I think I might've found a better way, but it's a little risky because it involves inheriting the Exceptions class and overwriting it. 

![image](https://user-images.githubusercontent.com/58575113/88629362-8299c500-d0f2-11ea-93dc-3e8fe7693a4d.png)

hallgchris replied:
What's the difference in the second method between that and just using the built-in `Exception` class? The first thing you suggested looks pretty neat and was what I was thinking of :)

khanguslee commented on overlay.py:
```diff
@@ -62,26 +69,36 @@ def publish_recording_status(self, message: str) -> None:
+		except:
+			self.client.loop_start()
+
+			while True:
```
Why is this in an infinite loop?

nix-chu replied:
This was the only way I could receive errors. Not sure why, but the script ending early was not publishing any errors.

khanguslee commented on backend/backend.py:
```diff
@@ -72,6 +73,12 @@ def stop_video(self) -> None:
 
+    def send_camera_error(self) -> None:
+        """ Sends the most recent exception for camera to the MQTT errors topic"""
+        message = { "error": format_exc() }
```
Can we put the trace stack value into a different key? Maybe `"trace"`?

khanguslee commented on overlay.py:
```diff
@@ -54,6 +56,11 @@ def __init__(self, bike, width=1280, height=740, bg: str=None):
 		self.start_time = time.time()
 
+	def publish_errors(self, message: dict) -> None:
+		message['camera'] = self.device
```
While we're here, can we also add additional details here too? Although the trace is informative, it would be good to have an idea of the current state of the overlay when it made those errors. This will lead to a faster debugging time for the engineers.
- `self.backend_name`
- `bike_version`
- `self.bg_path` if it is set
- `configs`

We should also have a `message` key with an easy to read message so that any subscriber to the error topic can just do `error.message`

There may be more that I might be missing and we might not even need the above. Let me know what you think.

nix-chu replied:
What should the message say?

nix-chu replied:
Should configs be accessed using `read_configs()`? Because right now, it is not an instance variable nor class variable.

khanguslee replied:
> What should the message say?

Maybe try to get it via `e.message` See: https://www.semicolonworld.com/question/59435/how-to-get-exception-message-in-python-properly
if the exception doesn't have a `message` just print the whole trace?

>Should configs be accessed using read_configs()? Because right now, it is not an instance variable nor class variable.

I think `read_configs()` is fine.

nix-chu replied:
Was able to append the configs to the JSON message just fine. 

On the other hand, not sure if the link is outdated but doing `e.message` doesn't seem to work. In case you were curious, simply printing `e` is just fine. For example:
```
>>>try:
>>>    raise Exception("Test error")
>>>except Exception as error_message:
>>>    print(error_message)
Test error
```

hallgchris commented on tests/mqtt_error_test.py:
```diff
@@ -0,0 +1,34 @@
+import time
```
Also, although we do need to fix the package structure like we did in data-acquisition-system, moving to the root folder will be a temporary fix for the imports not working correctly.

nix-chu replied:
Ahh alright, I had no idea. Will move it right away.

Would `overlay_error_test.py` be more appropriate? `overlay_error.py` might be mistaken as an actual overlay intended for the bike.

khanguslee replied:
I think so? See what happens when you run `pytest`. If it doesn't run your script, then it's fine.

nix-chu commented on overlay.py:
```diff
@@ -62,26 +69,36 @@ def publish_recording_status(self, message: str) -> None:
-				self.backend.on_loop()
+					self.backend.on_loop()
+		except:
+			self.client.loop_start()
```
Hm for some reason, regardless whether I have `self.client.loop_start()` or not, I can't seem to receive errors in my topic. Might need to troubleshoot this problem during our meeting today

nix-chu commented on overlay.py:
```diff
@@ -62,26 +69,36 @@ def publish_recording_status(self, message: str) -> None:
+				message = { "error": format_exc() }
+				self.publish_errors(message)
+				print(format_exc())
+				time.sleep(10) # Gives MQTT client time to publish 
```
I used the `.wait_for_publish` method, but now the script gets stuck and I still don't get a publish. Not sure why this is happening either.

nix-chu commented:
Wouldn't say that I'm done with this PR yet. But I'm re-requesting reviews just for clarification on some things. 

Changes:
- Renamed test file to `overlay_error.py` as it kept getting caught by Pytest. 
- Cleaned up error handling across the backends

I will try making that error handling class that will hopefully make the error code cleaner. I'm currently just trying to figure out how its going to work out at the moment. 
- Added a parameter that waits for the message to be published when set to True

However, this is currently causing the script to be frozen (camera is not even turning on either). Not sure how to overcome this problem...

khanguslee commented:
> Renamed test file to overlay_error.py as it kept getting caught by Pytest.

Thanks for checking this! I would make some sort of documentation at the top of the file to state that this overlay is not to be used and/or why it can be used.

>Cleaned up error handling across the backends

🎉 🎉 🎉 

>However, this is currently causing the script to be frozen (camera is not even turning on either). Not sure how to overcome this problem...

@hallgchris can probably answer this.

hallgchris commented on backend/backend.py:
```diff
@@ -72,6 +91,13 @@ def stop_video(self) -> None:
         """ Stops displaying the video feed and releases any resources
             captured. """
 
+    def send_camera_error(self, wait_for_publish: bool = False) -> None:
+        """ Sends the most recent exception for camera to the MQTT errors topic"""
+        message = { "trace": format_exc() }
+        self.publish_errors_func(message, wait_for_publish)
+        print(format_exc())
+        # TODO: Replace format_exc() with a logger
```
This function does the exact same thing as you do in the `except` clause in `overlay.py` - maybe this code should just be moved inside `publish_errors` in `overlay.py`?

hallgchris commented on overlay.py:
```diff
@@ -54,6 +56,23 @@ def __init__(self, bike, width=1280, height=740, bg: str=None):
 		self.start_time = time.time()
 
+	def publish_errors(self, message: dict, wait_for_publish: bool) -> None:
+		""" Sends camera error messages to the MQTT errors topic. Setting the 
+			wait_for_publish argument to True will block the broker until the 
```
Trailing spaces on these two lines :stuck_out_tongue_winking_eye:

hallgchris commented on overlay.py:
```diff
@@ -62,26 +81,32 @@ def publish_recording_status(self, message: str) -> None:
-					prev_data_update = time.time()
+					# Update the data overlay only if we have waited enough time
+					if time.time() > prev_data_update + self.data_update_interval:
+						prev_data_update = time.time()
 
-					# Update the data overlay with latest information
-					self.update_data_layer()
-					self.backend.on_canvases_updated(self.data_canvas, self.message_canvas)
+						# Update the data overlay with latest information
+						self.update_data_layer()
+						self.backend.on_canvases_updated(self.data_canvas, self.message_canvas)
 
-				self.backend.on_loop()
+					self.backend.on_loop()
```
If either of `self.update_data_layer()` or `self.backend.on_loop()` throw an exception, although it will be caught by the big outer try-catch, the `while True` loop will still be exited causing meaning that raspicam will close and the rider will loose vision. Not ideal :grimacing: 

To fix this, I reckon add some more of your try-catches inside the functions, like you've done for `on_canvases_updated`

hallgchris commented on overlay.py:
```diff
@@ -125,8 +150,11 @@ def _on_connect(self, client, userdata, flags, rc):
+		try:
+			payload = msg.payload.decode("utf-8")
+			self.data.load_data(msg.topic, payload)
+		except:
```
Make sure that every `except` you've got is `except Exception` :)

hallgchris commented on topics.py:
```diff
@@ -37,6 +37,7 @@ class Camera(Topic):
     get_overlays = 'camera/get_overlays'
     set_overlay = 'camera/set_overlay'
     push_overlays = 'camera/push_overlays'
+    errors = 'camera/errors'
```
I reckon keep the `/v3/` prefix. We should fix the rest of those topics to also have the prefix in a future PR

hallgchris commented on overlay_error.py:
```diff
@@ -0,0 +1,28 @@
+        try: 
+            raise Exception("Test error")
+        except:
+            self.backend.send_camera_error(True) # Set to true to block the broker until error is published
```
`paho-mqtt` is not in a fully connected state until the `on_connect` callback has returned. It also cannot publish until it is fully connected. However, the `on_connect` callback is unable to return until the message is published. So, we won't publish until we're connected, we won't connect until we've published, so nothing ever happens. If you change `True` to `False` here then everything works fine :)

nix-chu replied:
Hm, for some reason messages are able to be published on your computer, whereas mine doesn't seem to publish. Do you have time to troubleshoot this sometime before next software meeting? @hallgchris

hallgchris replied:
Sure thing! Will organise via slack

nix-chu commented on backend/backend.py:
```diff
@@ -36,15 +37,33 @@ def start_video(self) -> None:
+    def _on_base_canvas_updated(self, base_canvas: Canvas) -> None:
         """ Updates the base canvas which is drawn on the video. Should be
             called whenever the canvas is updated. """
+        raise NotImplementedError("No instructions for base layer found. ")
```
@hallgchris Don't these methods need to be implemented for the backend?

hallgchris replied:
They do need to be implemented, but what I'm saying is that you don't need to `raise NotImplementedError` because it's actually impossible to create a subclass without implementing this function. This is because of `@abstractmethod`. Give it a try! Create a class inheriting from `Backend` without implementing anything and try to create an instance of it, python will throw a runtime error and not let you

nix-chu replied:
Ohhhh righto, I see what you mean now. My bad, that was a misunderstanding

hallgchris commented on overlay.py:
```diff
@@ -171,8 +169,16 @@ def on_connect(self, client, userdata, flags, rc):
-	@abstractmethod
 	def update_data_layer(self):
+		""" Catches any errors that occurs while the data layer is being
+			updated. """
```
I've realised that this documentation (and similar descriptions in `backend.py` don't fully describe what happens when you call this function. When you call `update_data_layer`, while yes it does catch any errors that occur, its actual purpose is still to update the data layer. For these comments I think you should keep what you already have, but beforehand keep a description of what actually happens when you call the function. You could copy/paste a lot of this from the original function's comments.

hallgchris commented on overlay.py:
```diff
@@ -81,32 +85,26 @@ def publish_recording_status(self, message: str) -> None:
 
-		try:
-			with BackendFactory.create(self.backend_name, self.width, self.height, self.publish_recording_status, self.publish_errors) as self.backend:
+		with BackendFactory.create(self.backend_name, self.width, self.height, self.publish_recording_status, self.publish_errors) as self.backend:
```
We should probably still have some sort of error handling on the initialisation of the backend.

nix-chu commented:
Key Changes and Notes: 
- Added a new file that handles error handling in Raspicam

Yes, the file is called `camera_error_handler.py`, but I named the class `CameraException` so that it would be easier to identify that it is a camera exception, similar to doing `raise Exception`. I named the file the way it is because I thought it would be easier to identify when read from the Raspicam folder. Let me know if this irks you in any way. 😅

- CameraException class accepts 2 parameters; client and backend (name)

I passed `client` as an argument so that the message can be published. 

However I ran into the problem when an error occurs in the backend (or even instantiating this class using `with`), that it doesn't have access to the variables `client`, `camera/device`, `backend_name` and `bg_path` from the overlay. So as a result I made the following changes to the backend files which I think should be pointed out before this PR gets approved:

-`client` is now passed into BackendFactory
-Each backend has its own `backend_name` (which may be considered hard-coding) 😕
-I removed `bg_path` from the error message, because I wasn't able to see a nice clean way of passing it in the backend. Unless someone else knows another way, please let me know
-As for `camera/device`, I accessed these from `read_configs()` which I don't think should be a problem except for efficiency. 

- Removed the old methods of publishing camera errors HOORAY! 🎉
- An attempt at documentation

Let me know if the documentation for `camera_error_handler.py` is sufficient. 

- Error testing finally works! 😄 

![image](https://user-images.githubusercontent.com/58575113/88819955-0abbbf80-d204-11ea-99b5-4454b1361bcf.png)

Have a great day everybody

hallgchris commented on camera_error_handler.py:
```diff
@@ -0,0 +1,34 @@
+"""
+Exception handler for the Raspicam 
```
Lots of nitpicks here, sorry! Trailing space on this line

hallgchris commented on camera_error_handler.py:
```diff
@@ -0,0 +1,34 @@
+from config import read_configs
+from topics import Camera
+
+class CameraException:
```
While I like your thinking, I think the name `CameraException` is misleading. All the classes I can think of that end in `Exception` (or `Error`, btw) are ones which can be raised. (This is the definition of an exception.) However, `CameraException` is not an exception that can be raised - rather it is an exception handler, because rather than being an error itself, it takes some error that has occured and deals with it. I reckon `CameraErrorHandler` would be a good name!

tl;dr we can't `raise CameraException` so exception it is not.

hallgchris commented on camera_error_handler.py:
```diff
@@ -0,0 +1,34 @@
+from topics import Camera
+
+class CameraException:
+    def __init__(self, client, backend:str):
```
Just for consistency with other parts of this repo (and I think it's standard too) we should have a space after `:`

hallgchris commented on camera_error_handler.py:
```diff
@@ -0,0 +1,34 @@
+    def __enter__(self):
+        return self
+
+    def __exit__(self, type, value, traceback):
```
`type` isn't a good variable name, as `type` is also a python built-in. By naming a variable this, we're unable to use the python built-in. Consider something like `exc_type` or `exception_type`.

hallgchris commented on camera_error_handler.py:
```diff
@@ -0,0 +1,34 @@
+        return self
+
+    def __exit__(self, type, value, traceback):
+        if type is Exception: 
```
More trailing whitespace here

hallgchris commented on camera_error_handler.py:
```diff
@@ -0,0 +1,34 @@
+            }
+
+            status_topic = f"{str(Camera.errors)}"
+            publish_result = self.client.publish(status_topic, dumps(message))
```
We don't ever use the variable `publish_result` so it would be better not to assign to anything

hallgchris commented on backend/backend.py:
```diff
@@ -38,32 +39,32 @@ def start_video(self) -> None:
+            Catches any errors that occurs while the base canvas is being
             updated. """
-        try:
+        with CameraException(self.client, self.backend_name):
```
Rather than creating a new `CameraException` object every time we need error handling, we could store and instance of the class and only need to create it once in `__init__`. i.e.
```python
# in __init__
self.exceptionHandler = CameraException(client, self.backend_name)
# in functions
with self.exceptionHandler:
    # code
```
I haven't tested this, but I think it should work. Also would mean we don't need to store a reference to the `client`.

nix-chu replied:
Wait. That's actually brilliant

hallgchris commented on backend/backend.py:
```diff
@@ -38,32 +39,32 @@ def start_video(self) -> None:
+            Catches any errors that occurs while the base canvas is being
             updated. """
-        try:
+        with CameraException(self.client, self.backend_name):
```
Another thing, rather than hardcoding `self.backend_name` I believe we can use `self.__class__.__name__` (see https://stackoverflow.com/questions/19335722/getting-name-of-subclass-from-superclass/19335909)

hallgchris commented on camera_error_handler.py:
```diff
@@ -0,0 +1,34 @@
+                "message": str(value)
+            }
+
+            status_topic = f"{str(Camera.errors)}"
```
I don't think we need the f-string here - `str(Camera.errors)` should be fine

nix-chu commented:
Small changes and a big change!

- Renamed error handler class from CameraException to CameraErrorHandler
- Assigned CameraErrorHandler to an instance variable and used the variable for the `with` statement (this is brilliant Chris). No longer needs to instantiate in the backend and have awkward variables!
- Formating and documentation changes

hallgchris commented on overlay_error.py:
```diff
@@ -24,7 +20,7 @@ def on_connect(self, client, userdata, flags, rc):
 
         # Forcing an exception to occur
-        with CameraException(self.client, self.backend_name):
+        with self.exception_handler:
```
One last thing! (sorry) Let's add exception handling by default for `on_connect` by moving this line to `_on_connect`.

hallgchris replied:
I'm sorry, not quite still... By moving the error handling to `overlay.py` we no longer need to do `with self.exception_handler` in this file

nix-chu commented:
Changes:
- Discovered pylint and made it sometimes happy. (I prefer tabs over spaces tyvm)
- Using imperative form for method documentation. Not a big fan of imperative form and prefer present everyday tense, but it is what it is

nix-chu commented:
@khanguslee When you can, can you accept changes? Can't merge without your approval

khanguslee commented:
>Discovered pylint and made it sometimes happy. (I prefer tabs over spaces tyvm)

I make my tabs convert into spaces ;) Tabs looks weird on different machines, whereas spaces look more or less the same.
